### PR TITLE
Resolve type names to catalog

### PR DESCRIPTION
**Pull request by sploiselle**
Supersedes #5074

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5122)

### CODE REVIEW

File: src/coord/src/catalog/storage.rs [hunk 1]
```diff
@@ -244,6 +244,39 @@ impl Connection {
...
+        let tx = self.inner.transaction()?;
+        let current_setting: Option<String> = tx
+            .query_row(
+                "SELECT value FROM settings WHERE name = 'catalog_content_version';",
+                params![],
+                |row| row.get(0),
+            )
+            .optional()?;
+        let version = match current_setting {
+            Some(v) => v.parse::<i64>().unwrap() as usize,
```
Comment (benesch) [inline src/coord/src/catalog/storage.rs:257]: Getting the value as a string and then parsing it as an `i64` is roundabout. You should be able to get the value directly as a `u32`, then `usize::cast_from`.

Comment (benesch) [reply src/coord/src/catalog/storage.rs:257]:
```diff
diff --git a/src/coord/src/catalog/storage.rs b/src/coord/src/catalog/storage.rs
index 9de5d071a..3b4c1bcc8 100644
--- a/src/coord/src/catalog/storage.rs
+++ b/src/coord/src/catalog/storage.rs
@@ -246,7 +246,7 @@ impl Connection {
 
     pub fn get_catalog_content_version(&mut self) -> Result<usize, Error> {
         let tx = self.inner.transaction()?;
-        let current_setting: Option<String> = tx
+        let current_setting: Option<u32> = tx
             .query_row(
                 "SELECT value FROM settings WHERE name = 'catalog_content_version';",
                 params![],
@@ -254,7 +254,7 @@ impl Connection {
             )
             .optional()?;
         let version = match current_setting {
-            Some(v) => v.parse::<i64>().unwrap() as usize,
+            Some(v) => v,
             None => {
                 tx.execute(
                     "INSERT INTO settings (name, value) VALUES ('catalog_content_version', 0);",
@@ -264,7 +264,7 @@ impl Connection {
             }
         };
         tx.commit()?;
-        Ok(version as usize)
+        Ok(usize::cast_from(version))
     }
 
     pub fn set_catalog_content_version(&mut self, new_version: usize) -> Result<(), Error> {
```

File: src/coord/src/catalog/storage.rs [hunk 2]
```diff
@@ -244,6 +244,39 @@ impl Connection {
...
+        };
+        tx.commit()?;
+        Ok(version as usize)
+    }
+
+    pub fn set_catalog_content_version(&mut self, new_version: usize) -> Result<(), Error> {
+        let tx = self.inner.transaction()?;
+        tx.execute(
+            "UPDATE settings SET value = ? WHERE name = 'catalog_content_version'",
+            params![new_version as i64],
```
Comment (benesch) [inline src/coord/src/catalog/storage.rs:274]: Casting like this is sketchy. (Yes, we'll have other problems if we have more than `i64::MAX` migrations, but it still stresses me out!) `u32::try_from(new_version)` is probably the way to go—`u32` for parity with the recommendation above.

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
Very excited about this! Thanks, @sploiselle!

**Review by sploiselle — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5122)* status: 0 of 29 files reviewed, 9 unresolved discussions (waiting on @benesch)

___
*[src/coord/src/catalog.rs, line 389 at r2](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENkY-r2-389:-MP7PI9320AEO_N9EzSY:b-rjmusy) ([raw file](https://github.com/materializeinc/materialize/blob/a81dd1acb05f83f2dd56ad7644049ad7a40cb3e7/src/coord/src/catalog.rs#L389)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ooh, looks like these can just be function pointers + you can use a const:

```catalog
const CATALOG_CONTENT_MIGRATIONS: &[fn(&mut Catalog) -> Result<(), Error>] = &[];
```
</blockquote></details>

Done. Didn't realize it was possible to use non-capturing closures as function pointers. Very nice.
___
*[src/coord/src/catalog.rs, line 399 at r2](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENkY-r2-399:-MP7PJKi9TNET3sSEU1_:b-896fix) ([raw file](https://github.com/materializeinc/materialize/blob/a81dd1acb05f83f2dd56ad7644049ad7a40cb3e7/src/coord/src/catalog.rs#L399)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

This totally makes sense, but I think in this specific case it's a smidge cleaner to just make `Catalog::open` take a reference to the config:

```
diff --git a/src/coord/src/catalog.rs b/src/coord/src/catalog.rs
index c5d0eec1f..9d91c1a5b 100644
--- a/src/coord/src/catalog.rs
+++ b/src/coord/src/catalog.rs
@@ -432,11 +432,9 @@ impl Catalog {
     ///
     /// Returns the catalog and a list of events that describe the initial state
     /// of the catalog.
-    pub fn open(config: Config) -> Result<(Catalog, Vec<Event>), Error> {
+    pub fn open(config: &Config) -> Result<(Catalog, Vec<Event>), Error> {
         let (storage, experimental_mode, cluster_id) = storage::Connection::open(&config)?;
 
-        let cc = config.clone();
-
         let mut catalog = Catalog {
             by_name: BTreeMap::new(),
             by_id: BTreeMap::new(),
@@ -450,7 +448,7 @@ impl Catalog {
                 nonce: rand::random(),
                 experimental_mode,
                 cluster_id,
-                cache_directory: config.cache_directory,
+                cache_directory: config.cache_directory.clone(),
                 build_info: config.build_info,
             },
         };
@@ -679,7 +677,7 @@ impl Catalog {
                 .set_catalog_content_version(catalog_content_version + 1)?;
             // Reopen catalog with new contents to ensure on-disk and in-memory
             // state remain synchronized.
-            Catalog::open(cc)
+            Catalog::open(config)
         } else {
             Ok((catalog, events))
         }
@@ -692,7 +690,7 @@ impl Catalog {
     /// [`Catalog::open`] with appropriately set configuration parameters
     /// instead.
     pub fn open_debug(path: &Path) -> Result<Catalog, anyhow::Error> {
-        let (catalog, _) = Self::open(Config {
+        let (catalog, _) = Self::open(&Config {
             path,
             enable_logging: true,
             experimental_mode: None,
diff --git a/src/coord/src/coord.rs b/src/coord/src/coord.rs
index f4342545e..1af7eb997 100644
--- a/src/coord/src/coord.rs
+++ b/src/coord/src/coord.rs
@@ -3016,7 +3016,7 @@ where
         };
 
         let path = data_directory.join("catalog");
-        let (catalog, initial_catalog_events) = Catalog::open(catalog::Config {
+        let (catalog, initial_catalog_events) = Catalog::open(&catalog::Config {
             path: &path,
             experimental_mode: Some(experimental_mode),
             enable_logging: logging.is_some(),

```
</blockquote></details>

Done.
___
*[src/coord/src/catalog/storage.rs, line 257 at r2](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENkw-r2-257:-MP7PiRNEWoufveGlp3t:b-5evihp) ([raw file](https://github.com/materializeinc/materialize/blob/a81dd1acb05f83f2dd56ad7644049ad7a40cb3e7/src/coord/src/catalog/storage.rs#L257)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

```diff
diff --git a/src/coord/src/catalog/storage.rs b/src/coord/src/catalog/storage.rs
index 9de5d071a..3b4c1bcc8 100644
--- a/src/coord/src/catalog/storage.rs
+++ b/src/coord/src/catalog/storage.rs
@@ -246,7 +246,7 @@ impl Connection {
 
     pub fn get_catalog_content_version(&mut self) -> Result<usize, Error> {
         let tx = self.inner.transaction()?;
-        let current_setting: Option<String> = tx
+        let current_setting: Option<u32> = tx
             .query_row(
                 "SELECT value FROM settings WHERE name = 'catalog_content_version';",
                 params![],
@@ -254,7 +254,7 @@ impl Connection {
             )
             .optional()?;
         let version = match current_setting {
-            Some(v) => v.parse::<i64>().unwrap() as usize,
+            Some(v) => v,
             None => {
                 tx.execute(
                     "INSERT INTO settings (name, value) VALUES ('catalog_content_version', 0);",
@@ -264,7 +264,7 @@ impl Connection {
             }
         };
         tx.commit()?;
-        Ok(version as usize)
+        Ok(usize::cast_from(version))
     }
 
     pub fn set_catalog_content_version(&mut self, new_version: usize) -> Result<(), Error> {
```
</blockquote></details>

Done, but had to cast value to int within sqlite; there's no auto String to u32 casting within `rusqlite` afaict
___
*[src/coord/src/catalog/storage.rs, line 274 at r2](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENkw-r2-274:-MP7U75ZEiKhy4qowBAJ:bw6iods) ([raw file](https://github.com/materializeinc/materialize/blob/a81dd1acb05f83f2dd56ad7644049ad7a40cb3e7/src/coord/src/catalog/storage.rs#L274)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Casting like this is sketchy. (Yes, we'll have other problems if we have more than `i64::MAX` migrations, but it still stresses me out!) `u32::try_from(new_version)` is probably the way to go—`u32` for parity with the recommendation above.
</blockquote></details>

Done and used an `expect` here rather than an error we're unlikely to ever ecnounter
___
*[src/sql/src/plan/query.rs, line 2612 at r1](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENkn-r1-2612:-MP7Kmlv-6MGEzRiaz1L:bj2nzfp) ([raw file](https://github.com/materializeinc/materialize/blob/7b78bea304c6ed51d81a56a6a08048bc1370d3b3/src/sql/src/plan/query.rs#L2612)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Seems weird that `BIGINT` is handled in the parser but `smallint` is handled here instead!
</blockquote></details>

`smallint` has special casing in symbiosis (i.e. because `smallint` != `int4`) so needs to be preserved on input, whereas `bigint` == `int8`
___
*[src/sql/src/plan/query.rs, line 2678 at r1](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENkn-r1-2678:-MP7LEb29LEI0Seb2CuW:b-896fix) ([raw file](https://github.com/materializeinc/materialize/blob/7b78bea304c6ed51d81a56a6a08048bc1370d3b3/src/sql/src/plan/query.rs#L2678)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Unnecessary `mut`, I think.
</blockquote></details>

Done.
___
*[src/sql/src/plan/query.rs, line 2694 at r1](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENkn-r1-2694:-MP7LKfiAhSESFM8Cwf6:b-896fix) ([raw file](https://github.com/materializeinc/materialize/blob/7b78bea304c6ed51d81a56a6a08048bc1370d3b3/src/sql/src/plan/query.rs#L2694)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

This is an odd way to call a trait method. I think you probably want something like this?

```rust
val_ranges.iter().map(|r| format!("[{}-{}]", r.0, r.1)).join(",")
```
</blockquote></details>

Done.
___
*[src/sql/src/plan/query.rs, line 2698 at r1](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENkn-r1-2698:-MP7LYt51vh4CBkg_1rw:bk4er6t) ([raw file](https://github.com/materializeinc/materialize/blob/7b78bea304c6ed51d81a56a6a08048bc1370d3b3/src/sql/src/plan/query.rs#L2698)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Nit: this code seems undecided on whether "type mod" (singular) refers to a single `u64` or a slice of `u64`. Per the Postgres source code it seems like `&[u64]` is "type mods" and `u64` is "type mod"?
</blockquote></details>

in pg:
```
sean=# SELECT 1::numeric(1,2,3);
ERROR:  invalid NUMERIC type modifier
```

Seems like the whole thing is the type mod; will plumb the name of the type mod element through to the error message and hopefully elide some of the noise here
___
*[test/sqllogictest/string.slt, line 749 at r1](https://reviewable.io/reviews/materializeinc/materialize/5122#-MP4WaGrOuAfh2oLENks-r1-749:-MP7KA4f2PLncEkFZlyu:bhsiomg) ([raw file](https://github.com/materializeinc/materialize/blob/7b78bea304c6ed51d81a56a6a08048bc1370d3b3/test/sqllogictest/string.slt#L749)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

The term "type mods" is something I think we should avoid exposing to users. At least, I don't think PostgreSQL exposes it:

```
benesch=> select 1::char(1000000000);
ERROR:  length for type char cannot exceed 10485760
```

Could we pass along the meaning of each mod that is getting validated ("length", "precision", "scale", what-have-you) for surfacing in error messages?
</blockquote></details>

pg does surface the idea of a type modifier in some contexts:

```
sean=# SELECT 1::int4(1,2,3);
ERROR:  type modifier is not allowed for type "int4"
```

That being said, glad to improve the error messaging

**Pull request merged**